src/libsyntax/ext/tt/macro_check.rs

```diff
@@ -0,0 +1,342 @@
+use crate::ext::tt::quoted::{KleeneToken, TokenTree};
+use crate::parse::token::TokenKind;
+use crate::parse::token::{DelimToken, Token};
+use crate::parse::ParseSess;
+use crate::symbol::{kw, sym};
+
+use rustc_data_structures::fx::FxHashMap;
+use syntax_pos::{symbol::Ident, Span};
+
+/// Information attached to a meta-variable binder in LHS.
+struct BinderInfo {
+    /// The span of the meta-variable in LHS.
+    span: Span,
+    /// The stack of Kleene operators (outermost first).
+    ops: Vec<KleeneToken>,
+}
+
+/// An environment of meta-variables to their binder information.
+type Binders = FxHashMap<Ident, BinderInfo>;
+
+/// The state at which we entered a macro definition in the RHS of another macro definition.
```

Comment:
Perhaps add:
"This includes the meta-variables and stack of Kleene operators from nested outer macros. In most cases, we will pass around a stack of `MacroState` representing arbitrarily deep nesting of macros to get to the macro we are now processing. The depth of repetition of an arbitrary binder in the current macro is the sum of lengths of `ops` for all of the outer macros' `MacroState`s."